src/main/scala/firrtl/passes/ConstProp.scala

```diff
@@ -234,33 +234,52 @@ object ConstProp extends Pass {
     case _ => r
   }
 
+  // Two pass process
+  // 1. Propagate constants in expressions and forward propagate references
+  // 2. Propagate references again for backwards reference (Wires)
+  // TODO Replacing all wires with nodes makes the second pass unnecessary
   @tailrec
   private def constPropModule(m: Module): Module = {
     var nPropagated = 0L
     val nodeMap = collection.mutable.HashMap[String, Expression]()
 
+    def backPropExpr(expr: Expression): Expression = {
+      val old = expr map backPropExpr
+      val propagated = old match {
+        case ref @ WRef(rname, _,_, MALE) if nodeMap.contains(rname) =>
+          constPropNodeRef(ref, nodeMap(rname))
+        case x => x
+      }
+      if (old ne propagated) {
+        nPropagated += 1
+      }
+      propagated
+    }
+    def backPropStmt(stmt: Statement): Statement = stmt map backPropStmt map backPropExpr
+
     def constPropExpression(e: Expression): Expression = {
       val old = e map constPropExpression
       val propagated = old match {
         case p: DoPrim => constPropPrim(p)
         case m: Mux => constPropMux(m)
-        case r: WRef if nodeMap contains r.name => constPropNodeRef(r, nodeMap(r.name))
+        case ref @ WRef(rname, _,_, MALE) if nodeMap.contains(rname) =>
+          constPropNodeRef(ref, nodeMap(rname))
         case x => x
       }
-      if (old ne propagated)
-        nPropagated += 1
       propagated
     }
 
     def constPropStmt(s: Statement): Statement = {
-      s match {
+      val stmtx = s map constPropStmt map constPropExpression
+      stmtx match {
         case x: DefNode => nodeMap(x.name) = x.value
```

Comment:
Yeah this PR changes from postorder to preorder traversal. As far as I can tell it's functionally equivalent but results in a vast performance improvement because something like
```
node x = and(and(and(x, UInt<1>(1))))
```
can be done in one iteration than in multiple as it had been before. This is fixing the same type of performance bug as in https://github.com/freechipsproject/firrtl/pull/543 (why is why I separated that part out into a separate commit)